File: chap08_intercellular/Llinas.m

%  -------------------------------------------------------------------
%
%   Program to solve the Llinas model of synaptic suppression.
%
%   For Chapter 8, Section 8.1.2 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%   Written by James Keener and James Sneyd
%
%  -------------------------------------------------------------------

function Llinas

clear all
close all
clc

set(0,                       ...
'defaultaxesfontsize', 20,   ...
'defaultaxeslinewidth', 2.0, ...
'defaultlinelinewidth', 2.0);

global k10 k2 z1 s0 PCa F   ci ce n Vss k1  ohss tsw FbyRT

k10=2; k2=1; z1=1; s0=100; PCa=0.00001; F=96490; R=8.315; Tp=300;
ci=0.00001; ce=40; n=5;
Vss=-70;
FbyRT = F/(R*Tp*1000);  % units of mV^(-1)
k1=k10*exp(FbyRT*z1*Vss);
ohss= k1/(k1+k2);
tsw = 2.5;

t = [0:.01:5];
Vlist = [-40,-10,20,50,120];  % voltage in units of mV

figure(1)
for j = 1:length(Vlist)
    V = Vlist(j);
    oh = prob(t,V);
    currents= fCa(V,oh);
    ICa=currents(1,:);

    plot(t,-ICa,'LineWidth',2)
    xlabel('time (ms)')
    ylabel('-I_{Ca}(pA/(\mum)^2)')
    legend('boxoff')
    hold on
    ICa_save(j,:) = ICa;            % for external plotting
end
legend('V=-40mV','V=-10mV','V=20mV','V=50mV','V=120mV')

%save('Llinas.mat','t','ICa_save')   % for external plotting

x = linspace(-70,70,1000);  % units mV
phi_x=2*FbyRT*x;
Jss=PCa*2*F*phi_x.*(ci-ce*exp(-phi_x))./(1-exp(-phi_x));
Oss=k10*exp(FbyRT*z1*x).*(1./(k10*exp(FbyRT*z1*x)+k2));
ICass=Jss.*(s0/n).*((Oss).^n);

figure(2)
[hax, h1, h2] = plotyy(x,ICass,x,Oss);
ylabel(hax(1),'I_{Ca}')
ylim(hax(2),[0,1])
xlim([-70,70])
xlabel('V (mV)')
hold on
plot(x,Jss,'--');

% ylabel('\hat{o}')
legend('boxoff')
legend('I_{Ca}','j','\delta','Location','northwest','fontsize',18)


%save('Llinas.mat','x','ICass','Jss','Oss',"-append")  % for external plotting

for j = 1:length(Vlist)
    V = Vlist(j);
    Vt=(t<tsw).*V+(t>=tsw).*Vss;
    oh = probde(t,V);
    ICa = fCa(Vt,oh');

    ICa_save2(j,:) = ICa;       % for external plotting
    oh_save2(j,:) = oh;          % for external plotting

    figure(3)
    semilogy(t,-ICa,'LineWidth',2)
    xlabel('time (ms)')
    ylabel('-I_{Ca}(pA/(\mum)^2)')
    hold on
    legend('boxoff')

    figure(4)
    plot(t,oh)
    hold on
    legend('boxoff')

    xlabel('time (ms)')
    ylabel('open probability')
end

%save('Llinas.mat','ICa_save2','oh_save2',"-append")   % for external plotting


figure(3)
legend('V=-40mV','V=-10mV','V=20mV','V=50mV','V=120mV')
hold off
figure(4)
legend('V=-40mV','V=-10mV','V=20mV','V=50mV','V=120mV')
hold off


end % of main

%%     Calculate oh analytically
function oh=prob(t,V)
global k10 k2 z1  ohss FbyRT

k1=k10*exp(FbyRT*z1*V);
oh=k1./(k1+k2)*(1-exp(-(k1+k2)*t))+ohss*exp(-(k1+k2)*t);

end

%%
function currents=fCa(V,oh)
global  s0 PCa F ci ce n FbyRT

phi=2*FbyRT*V;
jj=PCa*2*F.*phi.*(ci-ce.*exp(-phi))./(1-exp(-phi));
ICa =(s0/n).*jj.*(oh.^n);

currents = [ICa];
end

%%  Calculate oh using the differential equation
function oh=probde(t,V)
global k10 k2 z1  Vss tsw FbyRT k2

k1=k10*exp(FbyRT*z1*Vss);
ohss= k1/(k1+k2);
sinit=ohss;
tspan = t;

[T,S] = ode15s(@(t,x)rhs(t,x,V),tspan,sinit);
% warning:  for piecewise continuous de's DO NOT use ode45 !!
oh =S(:,1);

end

%%
function out=rhs(t,s,V)
global Vss k10 k2 z1 tsw FbyRT

Vt=(t<tsw).*V+(t>=tsw).*Vss;
k1=k10*exp(FbyRT*z1*Vt);

out = k1.*(1-s)-k2*s;
end

File: chap08_intercellular/Llinas.py
#   -------------------------------------------------------------------
# 
#    Program to solve the Llinas model of synaptic suppression.
# 
#    For Chapter 8, Section 8.1.2 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
#  
#    Written by James Keener and James Sneyd
#  
#   ------------------------------------------------------------------- 

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import odeint


# Calculate oh analytically
def prob(t, V):
    k1 = k10 * np.exp(FbyRT * z1 * V)
    oh = k1 / (k1 + k2) * (1 - np.exp(-(k1 + k2) * t)) + ohss * np.exp(-(k1 + k2) * t)
    return oh


# calculate the currents
def fCa(V, oh):
    phi = 2 * FbyRT * V
    jj = PCa * 2 * F * phi * (ci - ce * np.exp(-phi)) / (1 - np.exp(-phi))
    ICa = (s0 / n) * jj * (oh ** n)
    return ICa



# Calculate oh using the differential equation
def probde(t, V):
    k1 = k10 * np.exp(FbyRT * z1 * Vss)
    ohss = k1 / (k1 + k2)
    sinit = ohss
    tspan = t

    S = odeint(rhs, sinit, tspan, args=(V,))
    oh = S[:, 0]
    return oh


# Define the right-hand side of the differential equation
def rhs(s, t, V):
    Vt = np.where(t < tsw, V, Vss)
    k1 = k10 * np.exp(FbyRT * z1 * Vt)
    return k1 * (1 - s) - k2 * s


k10 = 2
k2 = 1
z1 = 1
s0 = 100
PCa = 0.00001
F = 96490
R = 8.315
Tp = 300
ci = 0.00001
ce = 40
n = 5
Vss = -70
FbyRT = F / (R * Tp * 1000)  # units of mV^(-1)
k1 = k10 * np.exp(FbyRT * z1 * Vss)
ohss = k1 / (k1 + k2)
tsw = 2.5

# Time vector
t = np.arange(0, 5.01, 0.01)

# Voltage list
Vlist = [-40, -10, 20, 50, 120]

# SOlution for different voltage step increases
plt.figure(1)
ICa_save = np.zeros((len(Vlist), len(t)))
for j, V in enumerate(Vlist):
    oh = prob(t, V)
    ICa = fCa(V, oh)
    
    plt.plot(t, -ICa, linewidth=2)
    plt.xlabel('time (ms)')
    plt.ylabel('-I_{Ca}(pA/(\mum)^2)')
    plt.legend(['V=-40mV', 'V=-10mV', 'V=20mV', 'V=50mV', 'V=120mV'])



# Steady-state curves
x = np.linspace(-70, 70, 1000)
phi_x = 2 * FbyRT * x
Jss = PCa * 2 * F * phi_x * (ci - ce * np.exp(-phi_x)) / (1 - np.exp(-phi_x))
Oss = k10 * np.exp(FbyRT * z1 * x) * (1 / (k10 * np.exp(FbyRT * z1 * x) + k2))
ICass = Jss * (s0 / n) * (Oss ** n)

# Plot the steady-state curves
plt.figure(2)
fig, ax1 = plt.subplots()
ax2 = ax1.twinx() 
ax1.plot(x, ICass, 'r', x, Jss,'b', linewidth=2)
ax2.plot(x, Oss, 'g', linewidth=2)
ax1.set_xlabel('V (mV)')
ax1.set_ylabel(r'$I_{Ca}$')
ax2.set_ylabel('open probability',color='green')
ax2.tick_params(axis ='y', colors = 'green')
ax2.spines['right'].set_color('green')
plt.legend()
plt.show()


# Plot the responses to a step increase in V followed by a step decrease
for j, V in enumerate(Vlist):
    Vt = np.where(t < tsw, V, Vss)
    oh = probde(t, V)
    ICa = fCa(Vt, oh)

    plt.figure(3)
    plt.semilogy(t, -ICa, linewidth=2)
    plt.xlabel('time (ms)')
    plt.ylabel(r'$-I_{Ca}$ (pA/($\mu$ m)$^2)$')
    plt.legend(['V=-40mV', 'V=-10mV', 'V=20mV', 'V=50mV', 'V=120mV'])

    plt.figure(4)
    plt.plot(t, oh)
    plt.xlabel('time (ms)')
    plt.ylabel('open probability')

plt.figure(3)
plt.ylim([0.1,1e4])
plt.legend(['V=-40mV', 'V=-10mV', 'V=20mV', 'V=50mV', 'V=120mV'])
plt.figure(4)
plt.legend(['V=-40mV', 'V=-10mV', 'V=20mV', 'V=50mV', 'V=120mV'])




